src/python/pants/backend/python/goals/lockfile.py

```diff
@@ -269,44 +277,85 @@ async def generate_lockfiles_goal(
     return GenerateLockfilesGoal(exit_code=0)
 
 
+class AmbiguousResolveNamesError(Exception):
+    def __init__(self, ambiguous_names: list[str]) -> None:
+        if len(ambiguous_names) == 1:
+            first_paragraph = (
+                "The same resolve name is used both for a tool lockfile and a lockfile for your "
+                f"own code, so it's ambiguous what the resolve refers to: {ambiguous_names[0]}"
+            )
+        else:
+            first_paragraph = (
+                "The same resolve names are used both for tool lockfiles and lockfiles for your "
+                f"own code, so it's ambiguous what the resolves refer to: {sorted(ambiguous_names)}"
+            )
+        super().__init__(
+            f"{first_paragraph}\n\n"
+            "To fix, please update `[python-setup].experimental_resolves_to_lockfiles` to use "
+            "different resolve names."
+        )
+
+
 class UnrecognizedResolveNamesError(Exception):
-    pass
+    def __init__(
+        self, unrecognized_resolve_names: list[str], all_valid_names: Iterable[str]
+    ) -> None:
+        # TODO(#12314): maybe implement "Did you mean?"
+        if len(unrecognized_resolve_names) == 1:
+            unrecognized_str = unrecognized_resolve_names[0]
+            name_description = "name"
+        else:
+            unrecognized_str = str(sorted(unrecognized_resolve_names))
+            name_description = "names"
+        super().__init__(
+            f"Unrecognized resolve {name_description} from the option "
+            f"`--generate-lockfiles-resolve`: {unrecognized_str}\n\n"
+            f"All valid resolve names: {sorted(all_valid_names)}"
+        )
 
 
-def determine_tool_sentinels_to_generate(
+def determine_resolves_to_generate(
+    all_user_resolves: dict[str, str],
     all_tool_sentinels: Iterable[type[PythonToolLockfileSentinel]],
     requested_resolve_names: Sequence[str],
-) -> list[type[PythonToolLockfileSentinel]]:
-    if not requested_resolve_names:
-        return list(all_tool_sentinels)
+) -> tuple[dict[str, str], list[type[PythonToolLockfileSentinel]]]:
+    """Apply the `--resolve` option to determine which resolves are specified.
 
+    Return a tuple of user_resolves, tool_lockfile_sentinels.
+    """
     resolve_names_to_sentinels = {
         sentinel.options_scope: sentinel for sentinel in all_tool_sentinels
     }
+
+    ambiguous_resolve_names = set(resolve_names_to_sentinels.keys()).intersection(
+        all_user_resolves.keys()
+    )
+    if ambiguous_resolve_names:
```

Comment:
Take or leave: using set conversions and intersections is pretty verbose, and you can achieve the same results with a genexp:

```suggestion
    if any(i in all_user_resolves for i in resolve_names_to_sentinels):
```

Reply:
Generally agreed, but we need to preserve which were ambiguous - not just the boolean of if any were or not.

Reply:
(I see you're reusing `ambiguous_resolve_names`. Try:

```python
 [i for i in resolve_names_to_sentinels if i in all_user_resolves]
```